### PR TITLE
Json structured logs

### PR DESCRIPTION
**Pull request by loocapro**
Related to: #5726.

I tried to resume all the functionalities that were a bit spread between `bin/reth/src/cli/mod.rs` and `crates/tracing/src/lib.rs` into a single component defined in `crates/tracing/src/tracer.rs`.

Basic usage for a json stdout tracer would be:

```rust
Tracer::new().with_stdout(LayerInfo::new(
    LogFormat::Json,
    "debug".to_string(),
    LevelFilter::INFO.into(),
    None,
));
```
If no layer is configured it defaults to:

```rust
impl Default for LayerInfo {
    /// Provides default values for `LayerInfo`.
    ///
    /// By default, it uses terminal format, INFO level filter,
    /// no additional filters, and no color configuration.
    fn default() -> Self {
        Self {
            format: LogFormat::Terminal,
            directive: LevelFilter::INFO.into(),
            filters: "debug".to_string(),
            color: None,
        }
    }
}
```

I am a bit confused on why we use `log.filter` and `log.verbosity` as cli args.

So i have included them both in the `LayerInfo`, but i feel like since a directive is a rule derived from a filter, I can probably get rid of one of them.

LFF (looking for feedback!) please.

CC: @shekhirin @mattsse

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
this looks pretty good.

defer to @shekhirin and @onbjerg for a closer look

**Comment by mattsse** [comment]
friendly bump here @shekhirin @onbjerg

File: crates/tracing/src/layers.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        let layer = format.apply(file_filter, None, Some(writer));
+        self.inner.push(layer);
+        Ok(guard)
+    }
+}
+
+/// Holds configuration information for file logging.
+///
+/// Contains details about the log file's path, name, size, and rotation strategy.
+
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:129]:
```suggestion
```

File: crates/tracing/src/layers.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    /// * `format` - The format for log messages.
+    /// * `filter` - Additional filter directives as a string.
+    /// * `file_info` - Information about the log file including path and rotation strategy.
+    ///
+    /// # Returns
+    /// An `eyre::Result<FileWorkerGuard>` representing the file logging worker.
+    pub(crate) fn file(
+        &mut self,
+        format: LogFormat,
+        filter: String,
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:113]: just for the consistency with other methods
```suggestion
        filter: &str,
```

File: crates/tracing/src/layers.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    ///
+    /// # Returns
+    /// A tuple containing the non-blocking writer and its associated worker guard.
+    fn create_log_writer(
+        &self,
+        log_dir: &Path,
+    ) -> (tracing_appender::non_blocking::NonBlocking, WorkerGuard) {
+        let (writer, guard) = tracing_appender::non_blocking(
+            RollingFileAppender::new(
+                log_dir.join(AsRef::<Path>::as_ref(&self.file_name)),
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:164]: I think it should work with just a reference
```suggestion
                log_dir.join(&self.file_name),
```

**Review by shekhirin — COMMENTED** [review]
This looks great, just nits. Next step is integrating this into the existing logging configuration, right?

**Comment by shekhirin** [comment]
> I am a bit confused on why we use log.filter and log.verbosity as cli args.

verbosity is an easy way to display all logs for a certain level without messing with `RUST_LOG`

**Comment by loocapro** [comment]
> This looks great, just nits. Next step is integrating this into the existing logging configuration, right?

Great! Yep, I am going to proceed then!

**Comment by loocapro** [comment]
Hi @shekhirin , I have managed to hook up all components.

I have also added a Tracer trait in order to standardise the init function and remove the `init_test_tracing` function and just have a TestTracer struct implementing the Tracer trait. If it's ok, i can proceed to replace every `init_test_tracing` with the `TestTracer` trait.

**Comment by loocapro** [comment]
Hi @shekhirin gently bump here. I have managed to sync the code with latest main.

I have removed the TestTracer and just left the init_test_tracing function since it required a bit of changes around a lot of tests, but tell me if you want i can just do it.

I am a bit worried with dealing with more merge conflicts, so help me out here 😊

**Comment by mattsse** [comment]
ah sorry about the conflicts, we should have included this much sooner.

ptal @shekhirin

File: crates/tracing/src/layers.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+/// Contains details about the log file's path, name, size, and rotation strategy.
+#[derive(Debug, Clone)]
+pub struct FileInfo {
+    dir: PathBuf,
+    file_name: String,
+    max_size_bytes: u64,
+    max_files: usize,
+}
+
+const RETH_LOG_FILE_NAME: &str = "reth.log";
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:109]: move to the top of the file

File: crates/tracing/src/layers.rs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+impl FileInfo {
+    /// Creates a new `FileInfo` instance.
+    pub fn new(dir: PathBuf, max_size_bytes: u64, max_files: usize) -> Self {
+        Self { dir, file_name: RETH_LOG_FILE_NAME.to_string(), max_size_bytes, max_files }
+    }
+    /// Creates the log directory if it doesn't exist.
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:116]:
```suggestion
    }

    /// Creates the log directory if it doesn't exist.
```

File: crates/tracing/src/layers.rs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    /// Creates a non-blocking writer for the log file.
+    ///
+    /// # Arguments
+    /// * `log_dir` - Reference to the log directory path.
+    ///
+    /// # Returns
+    /// A tuple containing the non-blocking writer and its associated worker guard.
+    fn create_log_writer(
+        &self,
+        log_dir: &Path,
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:137]: do we need to accept the directory argument here, even though it's always passed as the one constructed by `create_log_dir`? Maybe we can combine `create_log_dir` and `create_log_writer` into a single function to simplify the usage?

File: crates/tracing/src/layers.rs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+/// A worker guard returned by the file layer.
+///
+///  When a guard is dropped, all events currently in-memory are flushed to the log file this guard
+///  belongs to.
+pub type FileWorkerGuard = tracing_appender::non_blocking::WorkerGuard;
+
+///  A boxed tracing [Layer].
+pub(crate) type BoxedLayer<S> = Box<dyn Layer<S> + Send + Sync>;
```
Comment (shekhirin) [inline crates/tracing/src/layers.rs:158]: move to the top of the file

File: crates/tracing/src/lib.rs [hunk 8]
```diff
@@ -17,120 +44,181 @@
...
+        self.journald = Some(filter);
+        self
+    }
+
+    ///  Sets the file layer configuration and associated file info.
+    ///
+    ///  # Arguments
+    ///  * `config` - The `LayerInfo` to use for the file layer.
+    ///  * `file_info` - The `FileInfo` containing details about the log file.
+
```
Comment (shekhirin) [inline crates/tracing/src/lib.rs:108]:
```suggestion
```

File: crates/tracing/src/lib.rs [hunk 9]
```diff
@@ -17,120 +44,181 @@
...
+    pub fn new(
+        format: LogFormat,
+        filters: String,
+        directive: Directive,
+        color: Option<String>,
+    ) -> Self {
+        Self { format, directive, filters, color }
+    }
+}
+impl Default for LayerInfo {
```
Comment (shekhirin) [inline crates/tracing/src/lib.rs:153]:
```suggestion
}

impl Default for LayerInfo {
```

File: crates/tracing/src/lib.rs [hunk 10]
```diff
@@ -17,120 +44,181 @@
...
-    S: Subscriber,
-    for<'a> S: LookupSpan<'a>,
-{
-    Ok(tracing_journald::layer()?.with_filter(filter).boxed())
+/// The `Tracer` trait provides a standardized way to initialize logging configurations
+/// in an application. Implementations of this trait can specify different logging setups,
+/// such as standard output logging, file logging, journald logging, or custom logging
+/// configurations tailored for specific environments (like testing).
+pub trait Tracer {
+    /// Initialize the logging configuration.
```
Comment (shekhirin) [inline crates/tracing/src/lib.rs:175]: needs a comment about the return type and how to handle it

File: crates/tracing/src/lib.rs [hunk 11]
```diff
@@ -17,120 +44,181 @@
...
-where
-    S: Subscriber,
-    for<'a> S: LookupSpan<'a>,
-{
-    Ok(tracing_journald::layer()?.with_filter(filter).boxed())
+/// The `Tracer` trait provides a standardized way to initialize logging configurations
+/// in an application. Implementations of this trait can specify different logging setups,
+/// such as standard output logging, file logging, journald logging, or custom logging
+/// configurations tailored for specific environments (like testing).
+pub trait Tracer {
```
Comment (shekhirin) [inline crates/tracing/src/lib.rs:174]: this is nice. I like that we can have a separate tracer for tests. Let's do it in another PR, so we can merge this one first.

**Review by shekhirin — CHANGES_REQUESTED** [review]
some questions on the API and overall simplification

File: bin/reth/src/args/log_args.rs [hunk 12]
```diff
@@ -60,55 +62,50 @@ pub struct LogArgs {
...
-                    Ok(env_filter.add_directive(directive.parse()?))
-                })
-        };
-
-        // Create and add the journald layer if enabled
+    /// Creates a LayerInfo instance.
+    fn layer(&self, format: LogFormat, use_color: bool) -> LayerInfo {
+        LayerInfo::new(
+            format,
+            self.log_file_filter.clone(),
```
[the remark this answers is not in the document]
Comment (loocapro) [reply bin/reth/src/args/log_args.rs:75]: Hi @shekhirin thanks for catching this!

Currently the EnvFilter for stdout is configured by verbosity (cli arg) + rust_log (env var). 

While the EnvFilter for file and journald is configured from the filter cli args (stdout does not have a filter cli arg).

This was already like this before, should we normalise this?

Something like: each layer has its own filter from cli args (which default values?), and stdout configures it with also the verbosity cli arg?

**Comment by loocapro** [comment]
Hi @shekhirin ,

all review comments are resolved.

Keep in mind there is a slightly different behaviour than what it was before:

I have added `log_stdout_filter` which defaults to empty string. Before the stdout EnvFilter was configured from the `RUST_LOG` env var and the `verbosity` cli arg.

Also the verbosity cli arg is only being used by the stdout layer.

Let me know if we can finally merge this.

Thanks a lot!!

**Comment by mattsse** [comment]
good to merge @shekhirin ?

**Comment by shekhirin** [comment]
@mattsse testing rn

**Review by shekhirin — APPROVED** [review]
LGTM, works as expected! Thank you for addressing all the review comments 😄

**Comment by shekhirin** [comment]
We will also need to change the CLI documentation in the book, and integrate the tracer into tests with the new `RethTracer` trait.

**Comment by loocapro** [comment]
> LGTM, works as expected! Thank you for addressing all the review comments 😄

Thank you for your reviews!

> We will also need to change the CLI documentation in the book, and integrate the tracer into tests with the new `RethTracer` trait.

I can take care of it.

**Pull request merged**